crates/wasmtime/src/module/registry.rs

```diff
@@ -48,66 +43,91 @@ impl StoreFrameInfo {
-    pub fn register(&mut self, module: &Arc<CompiledModule>) {
-        let (start, end) = module.code().range();
+    /// Registers a new module with the registry.
+    pub fn register(&mut self, module: &Module) {
+        let compiled_module = module.compiled_module();
+        let (start, end) = compiled_module.code().range();
 
-        // Ignore modules with no code or finished functions
-        if start == end || module.finished_functions().is_empty() {
+        // Ignore modules with no code, finished functions, or if the module is already registered
+        if start == end || compiled_module.finished_functions().is_empty() {
             return;
         }
 
         // The module code range is exclusive for end, so make it inclusive as it
         // may be a valid PC value
         let end = end - 1;
 
+        if self.0.get(&end).is_some() {
+            return;
+        }
+
         // Assert that this module's code doesn't collide with any other registered modules
-        if let Some((_, prev)) = self.ranges.range(end..).next() {
+        if let Some((_, prev)) = self.0.range(end..).next() {
             assert!(prev.start > end);
         }
-        if let Some((prev_end, _)) = self.ranges.range(..=start).next_back() {
+
+        if let Some((prev_end, _)) = self.0.range(..=start).next_back() {
             assert!(*prev_end < start);
         }
 
-        let prev = self.ranges.insert(
+        let prev = self.0.insert(
             end,
-            ModuleFrameInfo {
+            RegisteredModule {
                 start,
-                module: module.clone(),
+                module: compiled_module.clone(),
+                signatures: module.shared_signatures().clone(),
             },
         );
         assert!(prev.is_none());
 
-        GLOBAL_INFO.lock().unwrap().register(start, end, module);
+        GLOBAL_MODULES.lock().unwrap().register(start, end, module);
+    }
+
+    /// Looks up a trampoline from a shared signature index.
+    ///
+    /// This will search all modules associated with the store for a suitable trampoline
+    /// given the shared signature index.
+    pub fn lookup_trampoline(&self, index: VMSharedSignatureIndex) -> Option<VMTrampoline> {
+        for (_, m) in &self.0 {
+            if let Some(trampoline) = m.signatures.trampolines.get(index) {
+                return Some(trampoline);
+            }
+        }
+
+        None
     }
 }
 
-impl Drop for StoreFrameInfo {
+impl Drop for ModuleRegistry {
     fn drop(&mut self) {
-        let mut info = GLOBAL_INFO.lock().unwrap();
-        for end in self.ranges.keys() {
+        let mut info = GLOBAL_MODULES.lock().unwrap();
+        for end in self.0.keys() {
             info.unregister(*end);
         }
     }
 }
 
-/// Represents a module's frame information.
-#[derive(Clone)]
-pub struct ModuleFrameInfo {
+struct RegisteredModule {
     start: usize,
     module: Arc<CompiledModule>,
+    signatures: Arc<ModuleSharedSignatures>,
```

Comment:
This is one of the things I was a little worried about with the signatures and ownership of the returned shared indices. In the middle of this code it's not clear to me (immediately at least) how this is valid for the lifetime of a `RegisteredModule`. I figured that the lifetime of indexes for a signature would be tied to a `Module` itself but the `Module` isn't persisted here and we just retain a signature map.

This might just be that I don't have the signature management all in my head though. I don't think this is a great place for a comment per se, but I'm curious what you think about this. Also, how does this ensure that the module's signatures' registrations outlive this `RegisteredModule`?

Reply:
I'm fine with changing `RegisteredModule` to store `Module` and just rely on that to get at the compiled module and signatures.

I think the original intention was to just keep alive the bare minimum for the registry's operation, as with module linking implemented, there's a bunch of stuff stored in `Module` geared towards instantiation that could be dropped while the module is still in use by a `Store` (i.e. no longer instantiatable, but an instance still exists for the module).  But if it's confusing ownerships and lifetimes, perhaps it's not worth it.

Reply:
Re: the question about how this ensures signatures outlive the `RegisteredModule`: only that the module's shared signatures are only unregistered on drop of `ModuleSharedSignatures` and the `Arc` references the only owner of that. So for it to drop: the referencing `ModuleInner` needs to drop and all stores where the module is registered need to drop; at that point the signatures are no longer needed.

Reply:
Nah I think it's a good idea to not store `Module` here so we can split it up as necessary.

I think I just got confused a bit as we have so many module wrappers and module bits and pieces I couldn't keep them all straight in my head. I realize now though that the lifetime here is obvious in that this is *the* point a store retains a module's signatures, and the drop of these signatures is what deregisters.

I think that this is probably fine to leave as-is, I just needed to sit and think and understand more about what was going on